<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>0.1 Dealing with confounders | 06c-Remove_confounders.utf8</title>
  <meta name="description" content="" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="0.1 Dealing with confounders | 06c-Remove_confounders.utf8" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="0.1 Dealing with confounders | 06c-Remove_confounders.utf8" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  


<script src="libs/header-attrs-2.7/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<!-- for Facebook -->  
<meta property="og:url" content="http://hemberg-lab.github.io/scRNA.seq.course/" />
<meta property="og:description" content="In this course we will be surveying the existing problems as well as the available computational and statistical frameworks available for the analysis of scRNA-seq. The course is taught through the University of Cambridge Bioinformatics training unit, but the material found on these pages is meant to be used for anyone interested in learning about computational analysis of scRNA-seq data." />
<meta property="og:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- for Twitter -->          
<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:title" content="Analysis of single-cell RNA-seq data" />
<meta name="twitter:description" content="In this course we will be surveying the existing problems as well as the available computational and statistical frameworks available for the analysis of scRNA-seq. The course is taught through the University of Cambridge Bioinformatics training unit, but the material found on these pages is meant to be used for anyone interested in learning about computational analysis of scRNA-seq data." />
<meta name="twitter:image" content="http://hemberg-lab.github.io/scRNA.seq.course/figures/RNA-Seq_workflow-5.pdf.jpg" />

<!-- Google Analytics -->
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-71525309-1"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-71525309-1');
</script>

<!-- For Logo -->
<script>
  $(document).ready(function(){
    $(".book-header").prepend('<img id="SangerLogo" src="figures/sanger-logo-header.png" />')
  }
</script>

<script>
  $(document).ready(function(){
    $('.book-header').wrap( $('<a>').attr('href', 'https://www.sanger.ac.uk/').attr('target', '_blank'))
  });
</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="index.html">Table of Contents</a></li>

<li class="divider"></li>
<li class="chapter" data-level="0.1" data-path=""><a href="#dealing-with-confounders"><i class="fa fa-check"></i><b>0.1</b> Dealing with confounders</a>
<ul>
<li class="chapter" data-level="0.1.1" data-path=""><a href="#introduction"><i class="fa fa-check"></i><b>0.1.1</b> Introduction</a></li>
<li class="chapter" data-level="0.1.2" data-path=""><a href="#combat"><i class="fa fa-check"></i><b>0.1.2</b> Combat</a></li>
<li class="chapter" data-level="0.1.3" data-path=""><a href="#mnncorrect-batchelor"><i class="fa fa-check"></i><b>0.1.3</b> mnnCorrect (batchelor)</a></li>
<li class="chapter" data-level="0.1.4" data-path=""><a href="#how-to-evaluate-and-compare-confounder-removal-strategies"><i class="fa fa-check"></i><b>0.1.4</b> How to evaluate and compare confounder removal strategies</a></li>
<li class="chapter" data-level="0.1.5" data-path=""><a href="#big-exercise"><i class="fa fa-check"></i><b>0.1.5</b> Big Exercise</a></li>
<li class="chapter" data-level="0.1.6" data-path=""><a href="#sessioninfo"><i class="fa fa-check"></i><b>0.1.6</b> sessionInfo()</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="http://www.sanger.ac.uk/science/groups/hemberg-group" target="blank">Hemberg Lab</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<!--bookdown:title:end-->
<!--bookdown:title:start-->
<div id="dealing-with-confounders" class="section level2" number="0.1">
<h2><span class="header-section-number">0.1</span> Dealing with confounders</h2>
<div id="introduction" class="section level3" number="0.1.1">
<h3><span class="header-section-number">0.1.1</span> Introduction</h3>
<p>In the previous chapter we normalized for library size, effectively removing it as a confounder. Now we will consider removing other less well-defined confounders from our data. Technical confounders (aka batch effects) can arise from difference in reagents, isolation methods, the lab/experimenter who performed the experiment, even which day or time of the day the experiment was performed. Accounting for technical confounders, and batch effects particularly, is a large topic that also involves principles of experimental design. Here we address approaches that can be taken to account for confounders when the experimental design is appropriate.</p>
<p>Fundamentally, accounting for technical confounders involves identifying and, ideally, removing sources of variation in the expression data that are not related to (i.e. are confounding) the biological signal of interest. Various approaches exist, some of which use spike-in or housekeeping genes, and some of which use endogenous genes.</p>
<div id="advantages-and-disadvantages-of-using-spike-ins-to-remove-confounders" class="section level4" number="0.1.1.1">
<h4><span class="header-section-number">0.1.1.1</span> Advantages and disadvantages of using spike-ins to remove confounders</h4>
<p>The use of spike-ins as control genes is appealing, since the same amount of ERCC (or other) spike-in was added to each cell in our experiment. In principle, all the variablity we observe for these genes is due to technical noise; whereas endogenous genes are affected by both technical noise and biological variability. Technical noise can be removed by fitting a model to the spike-ins and “substracting” this from the endogenous genes. There are several methods available based on this premise (eg. <a href="https://github.com/catavallejos/BASiCS">BASiCS</a>, <a href="https://github.com/PMBio/scLVM">scLVM</a>, <a href="http://bioconductor.org/packages/release/bioc/html/RUVSeq.html">RUVg</a>); each using different noise models and different fitting procedures. Alternatively, one can identify genes which exhibit significant variation beyond technical noise (eg. Distance to median, <a href="http://www.nature.com/nmeth/journal/v10/n11/full/nmeth.2645.html">Highly variable genes</a>). However, there are issues with the use of spike-ins for normalisation (particularly ERCCs, derived from bacterial sequences), including that their variability can, for various reasons, actually be <em>higher</em> than that of endogenous genes.</p>
<p>Given the issues with using spike-ins, better results can often be obtained by using endogenous genes instead. Where we have a large number of endogenous genes that, on average, do not vary systematically between cells and where we expect technical effects to affect a large number of genes (a very common and reasonable assumption), then such methods (for example, the RUVs method) can perform well.</p>
<p>We explore both general approaches below.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scRNA.seq.funcs)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SingleCellExperiment)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scran)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sva)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(batchelor)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(kBET)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1234567</span>)</span></code></pre></div>
<p>Things are different today, I hear every mother say! (The Rolling Stones about SCE object updates)</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>umi    <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;data/tung/umi.rds&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>umi.qc <span class="ot">&lt;-</span> umi[<span class="fu">rowData</span>(umi)<span class="sc">$</span>use, <span class="sc">!</span> <span class="fu">colData</span>(umi)<span class="sc">$</span>discard]</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>erccs  <span class="ot">&lt;-</span> <span class="fu">rownames</span>(<span class="fu">altExp</span>(umi.qc))</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>qclust <span class="ot">&lt;-</span> <span class="fu">quickCluster</span>(umi.qc, <span class="at">min.size =</span> <span class="dv">30</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>umi.qc <span class="ot">&lt;-</span> <span class="fu">computeSumFactors</span>(umi.qc, <span class="at">sizes =</span> <span class="dv">15</span>, <span class="at">clusters =</span> qclust)</span></code></pre></div>
<pre><code>## Warning in .guess_min_mean(x, min.mean = min.mean, BPPARAM = BPPARAM): assuming
## UMI data when setting &#39;min.mean&#39;</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>umi.qc <span class="ot">&lt;-</span> <span class="fu">logNormCounts</span>(umi.qc)</span></code></pre></div>
</div>
</div>
<div id="combat" class="section level3" number="0.1.2">
<h3><span class="header-section-number">0.1.2</span> Combat</h3>
<p>If you have an experiment with a balanced design, <code>Combat</code> can be used to eliminate batch effects while preserving biological effects by specifying the biological effects using the <code>mod</code> parameter. However the <code>Tung</code> data contains multiple experimental replicates rather than a balanced design so using <code>mod1</code> to preserve biological variability will result in an error.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>combat_data <span class="ot">&lt;-</span> <span class="fu">logcounts</span>(umi.qc)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>mod_data <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(combat_data))</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Basic batch removal</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>mod0 <span class="ot">=</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> <span class="dv">1</span>, <span class="at">data =</span> mod_data) </span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Preserve biological variability</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>mod1 <span class="ot">=</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> umi.qc<span class="sc">$</span>individual, <span class="at">data =</span> mod_data) </span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co"># adjust for total genes detected</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>mod2 <span class="ot">=</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> umi.qc<span class="sc">$</span>detected, <span class="at">data =</span> mod_data)</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(umi.qc, <span class="st">&quot;combat&quot;</span>) <span class="ot">&lt;-</span> <span class="fu">ComBat</span>(</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">dat =</span> <span class="fu">t</span>(mod_data), </span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">batch =</span> <span class="fu">factor</span>(umi.qc<span class="sc">$</span>batch), </span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">mod =</span> mod0,</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">par.prior =</span> <span class="cn">TRUE</span>,</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">prior.plots =</span> <span class="cn">FALSE</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>## Found 2775 genes with uniform expression within a single batch (all zeros); these will not be adjusted for batch.</code></pre>
<p><strong>Exercise 1</strong></p>
<p>Perform <code>ComBat</code> correction accounting for total features as a co-variate. Store the corrected matrix in the <code>combat_tf</code> slot.</p>
<pre><code>## Found 2775 genes with uniform expression within a single batch (all zeros); these will not be adjusted for batch.</code></pre>
</div>
<div id="mnncorrect-batchelor" class="section level3" number="0.1.3">
<h3><span class="header-section-number">0.1.3</span> mnnCorrect (batchelor)</h3>
<p><code>mnnCorrect</code> <span class="citation">[@Haghverdi2017-vh]</span> assumes that each batch shares at least one biological condition with each other batch. Thus it works well for a variety of balanced experimental designs. However, the <code>Tung</code> data contains multiple replicates for each invidividual rather than balanced batches, thus we will normalized each individual separately. Note that this will remove batch effects between batches within the same individual but not the batch effects between batches in different individuals, due to the confounded experimental design.</p>
<p>Thus we will merge a replicate from each individual to form three batches.</p>
<p>The paragraph above is not relevant anymore; mnnCorrect is only included in <code>batchelor</code> as a legacy function. It is now recommended to use <code>fastMNN</code> for MNN-based batch correction – AP.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>mnn_out <span class="ot">&lt;-</span> <span class="fu">fastMNN</span>(umi.qc,<span class="at">batch =</span> umi.qc<span class="sc">$</span>individual)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(umi.qc, <span class="st">&quot;mnn&quot;</span>) <span class="ot">&lt;-</span> <span class="fu">assay</span>(mnn_out,<span class="st">&#39;reconstructed&#39;</span>)</span></code></pre></div>
</div>
<div id="how-to-evaluate-and-compare-confounder-removal-strategies" class="section level3" number="0.1.4">
<h3><span class="header-section-number">0.1.4</span> How to evaluate and compare confounder removal strategies</h3>
<p>A key question when considering the different methods for removing confounders is how to quantitatively determine which one is the most effective. The main reason why comparisons are challenging is because it is often difficult to know what corresponds to technical counfounders and what is interesting biological variability. Here, we consider three different metrics which are all reasonable based on our knowledge of the experimental design. Depending on the biological question that you wish to address, it is important to choose a metric that allows you to evaluate the confounders that are likely to be the biggest concern for the given situation.</p>
<div id="effectiveness-1" class="section level4" number="0.1.4.1">
<h4><span class="header-section-number">0.1.4.1</span> Effectiveness 1</h4>
<p>We evaluate the effectiveness of the normalization by inspecting the
PCA plot where colour corresponds the technical replicates and shape
corresponds to different biological samples (individuals). Separation of biological samples and
interspersed batches indicates that technical variation has been
removed. We always use log2-cpm normalized data to match the assumptions of PCA.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(n <span class="cf">in</span> <span class="fu">assayNames</span>(umi.qc)) {</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>    tmp <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(umi.qc, <span class="at">exprs_values =</span> n, <span class="at">ncomponents =</span> <span class="dv">20</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">plotPCA</span>(</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>            tmp,</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>            <span class="at">colour_by =</span> <span class="st">&quot;batch&quot;</span>,</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>            <span class="at">size_by =</span> <span class="st">&quot;detected&quot;</span>,</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>            <span class="at">shape_by =</span> <span class="st">&quot;individual&quot;</span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>        ) <span class="sc">+</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ggtitle</span>(n)</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p><img src="06c-Remove_confounders_files/figure-html/remove-conf11-1.png" width="672" style="display: block; margin: auto;" /><img src="06c-Remove_confounders_files/figure-html/remove-conf11-2.png" width="672" style="display: block; margin: auto;" /><img src="06c-Remove_confounders_files/figure-html/remove-conf11-3.png" width="672" style="display: block; margin: auto;" /><img src="06c-Remove_confounders_files/figure-html/remove-conf11-4.png" width="672" style="display: block; margin: auto;" /><img src="06c-Remove_confounders_files/figure-html/remove-conf11-5.png" width="672" style="display: block; margin: auto;" /><img src="06c-Remove_confounders_files/figure-html/remove-conf11-6.png" width="672" style="display: block; margin: auto;" /></p>
<p><strong>Exercise 3</strong></p>
<p>Consider different <code>ks</code> for RUV normalizations. Which gives the best results?</p>
</div>
<div id="effectiveness-2" class="section level4" number="0.1.4.2">
<h4><span class="header-section-number">0.1.4.2</span> Effectiveness 2</h4>
<p>We can also examine the effectiveness of correction using the relative log expression (RLE) across cells to confirm technical noise has been removed from the dataset. Note RLE only evaluates whether the number of genes higher and lower than average are equal for each cell - i.e. systemic technical effects. Random technical noise between batches may not be detected by RLE.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(n <span class="cf">in</span> <span class="fu">assayNames</span>(umi.qc)) {</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>    res[[n]] <span class="ot">&lt;-</span> <span class="fu">suppressWarnings</span>(<span class="fu">calc_cell_RLE</span>(<span class="fu">assay</span>(umi.qc, n)))</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar=</span><span class="fu">c</span>(<span class="dv">6</span>,<span class="dv">4</span>,<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(res, <span class="at">las=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="06c-Remove_confounders_files/figure-html/remove-conf12-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="effectiveness-3" class="section level4" number="0.1.4.3">
<h4><span class="header-section-number">0.1.4.3</span> Effectiveness 3</h4>
<p>Another method to check the efficacy of batch-effect correction is to consider the intermingling of points from different batches in local subsamples of the data. If there are no batch-effects then proportion of cells from each batch in any local region should be equal to the global proportion of cells in each batch.</p>
<p><code>kBET</code> <span class="citation">[@Buttner2017-ds]</span> takes <code>kNN</code> networks around random cells and tests the number of cells from each batch against a binomial distribution. The rejection rate of these tests indicates the severity of batch-effects still present in the data (high rejection rate = strong batch effects). <code>kBET</code> assumes each batch contains the same complement of biological groups, thus it can only be applied to the entire dataset if a perfectly balanced design has been used. However, <code>kBET</code> can also be applied to replicate-data if it is applied to each biological group separately. In the case of the Tung data, we will apply <code>kBET</code> to each individual independently to check for residual batch effects. However, this method will not identify residual batch-effects which are confounded with biological conditions. In addition, <code>kBET</code> does not determine if biological signal has been preserved.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>compare_kBET_results <span class="ot">&lt;-</span> <span class="cf">function</span>(sce){</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>    indiv <span class="ot">&lt;-</span> <span class="fu">unique</span>(sce<span class="sc">$</span>individual)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>    norms <span class="ot">&lt;-</span> <span class="fu">assayNames</span>(sce) <span class="co"># Get all normalizations</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    results <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (i <span class="cf">in</span> indiv){ </span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> (j <span class="cf">in</span> norms){</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>            tmp <span class="ot">&lt;-</span> <span class="fu">kBET</span>(</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>                <span class="at">df =</span> <span class="fu">t</span>(<span class="fu">assay</span>(sce[,sce<span class="sc">$</span>individual<span class="sc">==</span> i], j)), </span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">batch =</span> sce<span class="sc">$</span>batch[sce<span class="sc">$</span>individual<span class="sc">==</span>i], </span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>                <span class="at">heuristic =</span> <span class="cn">TRUE</span>, </span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>                <span class="at">verbose =</span> <span class="cn">FALSE</span>, </span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>                <span class="at">addTest =</span> <span class="cn">FALSE</span>, </span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>                <span class="at">plot =</span> <span class="cn">FALSE</span>)</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>            results[[i]][[j]] <span class="ot">&lt;-</span> tmp<span class="sc">$</span>summary<span class="sc">$</span>kBET.observed[<span class="dv">1</span>]</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(<span class="fu">as.data.frame</span>(results))</span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a>eff_debatching <span class="ot">&lt;-</span> <span class="fu">compare_kBET_results</span>(umi.qc)</span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;reshape2&quot;</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;RColorBrewer&quot;</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot results</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>dod <span class="ot">&lt;-</span> <span class="fu">melt</span>(<span class="fu">as.matrix</span>(eff_debatching),  <span class="at">value.name =</span> <span class="st">&quot;kBET&quot;</span>)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(dod)[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Normalisation&quot;</span>, <span class="st">&quot;Individual&quot;</span>)</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>colorset <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&#39;gray&#39;</span>, <span class="fu">brewer.pal</span>(<span class="at">n =</span> <span class="dv">9</span>, <span class="st">&quot;RdYlBu&quot;</span>))</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(dod, <span class="fu">aes</span>(Normalisation, Individual, <span class="at">fill=</span>kBET)) <span class="sc">+</span>  </span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_tile</span>() <span class="sc">+</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_gradient2</span>(</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">na.value =</span> <span class="st">&quot;gray&quot;</span>,</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">low =</span> colorset[<span class="dv">2</span>],</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">mid=</span>colorset[<span class="dv">6</span>],</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>        <span class="at">high =</span> colorset[<span class="dv">10</span>],</span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>        <span class="at">midpoint =</span> <span class="fl">0.5</span>, <span class="at">limit =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_discrete</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_discrete</span>(<span class="at">expand =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>)) <span class="sc">+</span> </span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(</span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(</span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>            <span class="at">angle =</span> <span class="dv">45</span>, </span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>            <span class="at">vjust =</span> <span class="dv">1</span>, </span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>            <span class="at">size =</span> <span class="dv">12</span>, </span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>            <span class="at">hjust =</span> <span class="dv">1</span></span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>    ) <span class="sc">+</span> </span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">&quot;Effect of batch regression methods per individual&quot;</span>)</span></code></pre></div>
<p><img src="06c-Remove_confounders_files/figure-html/remove-conf14-1.png" width="672" style="display: block; margin: auto;" /></p>
<p><strong>Exercise 4</strong></p>
<p>Why do the raw counts appear to have little batch effects?</p>
</div>
</div>
<div id="big-exercise" class="section level3" number="0.1.5">
<h3><span class="header-section-number">0.1.5</span> Big Exercise</h3>
<p>Perform the same analysis with read counts of the <code>tung</code> data. Use <code>tung/reads.rds</code> file to load the reads <code>SCE</code> object. Once you have finished please compare your results to ours (next chapter). Additionally, experiment with other combinations of normalizations and compare the results.</p>
</div>
<div id="sessioninfo" class="section level3" number="0.1.6">
<h3><span class="header-section-number">0.1.6</span> sessionInfo()</h3>
<pre><code>## R version 4.0.4 (2021-02-15)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] RColorBrewer_1.1-2          reshape2_1.4.4             
##  [3] kBET_0.99.6                 batchelor_1.6.2            
##  [5] sva_3.38.0                  BiocParallel_1.24.1        
##  [7] genefilter_1.72.1           mgcv_1.8-34                
##  [9] nlme_3.1-152                scran_1.18.5               
## [11] scater_1.18.6               ggplot2_3.3.4              
## [13] SingleCellExperiment_1.12.0 SummarizedExperiment_1.20.0
## [15] Biobase_2.50.0              GenomicRanges_1.42.0       
## [17] GenomeInfoDb_1.26.4         IRanges_2.24.1             
## [19] S4Vectors_0.28.1            BiocGenerics_0.36.0        
## [21] MatrixGenerics_1.2.1        matrixStats_0.58.0         
## [23] scRNA.seq.funcs_0.1.0       knitr_1.31                 
## 
## loaded via a namespace (and not attached):
##   [1] Rtsne_0.15                ggbeeswarm_0.6.0         
##   [3] colorspace_2.0-0          ellipsis_0.3.1           
##   [5] scuttle_1.0.4             bluster_1.0.0            
##   [7] XVector_0.30.0            BiocNeighbors_1.8.2      
##   [9] farver_2.1.0              bit64_4.0.5              
##  [11] AnnotationDbi_1.52.0      fansi_0.4.2              
##  [13] codetools_0.2-18          splines_4.0.4            
##  [15] sparseMatrixStats_1.2.1   cachem_1.0.4             
##  [17] jsonlite_1.7.2            ResidualMatrix_1.0.0     
##  [19] annotate_1.68.0           cluster_2.1.1            
##  [21] compiler_4.0.4            httr_1.4.2               
##  [23] dqrng_0.2.1               assertthat_0.2.1         
##  [25] Matrix_1.3-2              fastmap_1.1.0            
##  [27] limma_3.46.0              BiocSingular_1.6.0       
##  [29] htmltools_0.5.1.1         tools_4.0.4              
##  [31] rsvd_1.0.3                igraph_1.2.6             
##  [33] gtable_0.3.0              glue_1.4.2               
##  [35] GenomeInfoDbData_1.2.4    dplyr_1.0.5              
##  [37] Rcpp_1.0.6                jquerylib_0.1.3          
##  [39] vctrs_0.3.6               DelayedMatrixStats_1.12.3
##  [41] xfun_0.22                 stringr_1.4.0            
##  [43] beachmat_2.6.4            lifecycle_1.0.0          
##  [45] irlba_2.3.3               hypergeo_1.2-13          
##  [47] statmod_1.4.35            XML_3.99-0.6             
##  [49] edgeR_3.32.1              zlibbioc_1.36.0          
##  [51] MASS_7.3-53.1             scales_1.1.1             
##  [53] yaml_2.2.1                memoise_2.0.0            
##  [55] gridExtra_2.3             sass_0.3.1               
##  [57] stringi_1.5.3             RSQLite_2.2.4            
##  [59] highr_0.8                 orthopolynom_1.0-5       
##  [61] contfrac_1.1-12           rlang_0.4.10             
##  [63] pkgconfig_2.0.3           moments_0.14             
##  [65] bitops_1.0-6              evaluate_0.14            
##  [67] lattice_0.20-41           purrr_0.3.4              
##  [69] labeling_0.4.2            cowplot_1.1.1            
##  [71] bit_4.0.4                 tidyselect_1.1.0         
##  [73] deSolve_1.28              plyr_1.8.6               
##  [75] magrittr_2.0.1            bookdown_0.22            
##  [77] R6_2.5.0                  generics_0.1.0           
##  [79] DelayedArray_0.16.3       DBI_1.1.1                
##  [81] pillar_1.5.1              withr_2.4.1              
##  [83] survival_3.2-10           RCurl_1.98-1.3           
##  [85] tibble_3.1.0              crayon_1.4.1             
##  [87] utf8_1.2.1                rmarkdown_2.7            
##  [89] viridis_0.5.1             locfit_1.5-9.4           
##  [91] grid_4.0.4                FNN_1.1.3                
##  [93] blob_1.2.1                digest_0.6.27            
##  [95] xtable_1.8-4              elliptic_1.4-0           
##  [97] munsell_0.5.0             beeswarm_0.3.1           
##  [99] viridisLite_0.3.0         vipor_0.4.5              
## [101] bslib_0.2.4</code></pre>
</div>
</div>
            </section>

          </div>
        </div>
      </div>


    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "section"
},
"search": false
});
});
</script>

</body>

</html>
